include/builtins.hpp

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <cstdio>
+#include <type_traits>
+/*
+template<class T> T __sync_fetch_and_add(T* ptr, T val) {
+   T old = *ptr;
+   *ptr += typename std::conditional<std::is_pointer<T>::value, size_t, T>::type( val );
+   return old;
+}
+
+template<class T> T __sync_add_and_fetch(T* ptr, T val) {
+   return ptr += typename std::conditional<std::is_pointer<T>::value, size_t, T>::type( val );
+}
+
+template<class T> T __sync_lock_test_and_set(T* ptr, T val) {
+   T old = *ptr;
+   *ptr = typename std::conditional<std::is_pointer<T>::value, size_t, T>::type( val );
+   return old;
+}
+*/
+
+inline int32_t __sync_lock_test_and_set( int32_t* ptr, int32_t val ) {
+   auto old = *ptr;
+   *ptr = val;
+   return old;
+}
+
+inline int32_t __sync_fetch_and_add( int32_t* ptr, int32_t val ) {
+   auto old = *ptr;
+   *ptr += val;
+   return old;
+}
+
+inline int32_t __sync_add_and_fetch( int32_t* ptr, int32_t val ) {
+   return *ptr += val;
+}
```

Comment:
Why the different style?

Reply:
@gleehokie , I am not sure that I understand what do you mean.
If your comment is about  __sync_fetch_and_add vs  __sync_fetch_and_add - they are different functions (first one returns old value while the second returns new value). libc++ needs them both in order to compile.